pkg/types/types.go

```diff
@@ -30,19 +33,60 @@ const (
 	KubeEndPoint = "kubelet.sock"
 )
 
-// ResourceConfig contains cofiguration paremeters for a resource pool
+// DeviceType is custom type to define supported device types
+type DeviceType string
+
+const (
+	// NetDeviceType is DeviceType for network class devices
+	NetDeviceType DeviceType = "netDevice"
+	// AcceleratorType is DeviceType for accelerator class devices
+	AcceleratorType DeviceType = "accelerator"
+)
+
+// SupportedDevices is map of 'device identifier as string' to 'device class hexcode as int'
+/*
+Network controller subclasses. ref: https://pci-ids.ucw.cz/read/PD/02
+00	Ethernet controller
+01	Token ring network controller
+02	FDDI network controller
+03	ATM network controller
+04	ISDN controller
+05	WorldFip controller
+06	PICMG controller
+07	Infiniband controller
+08	Fabric controller
+80	Network controller
+*/
+var SupportedDevices = map[DeviceType]int{
+	NetDeviceType:   0x02,
+	AcceleratorType: 0x12,
+}
+
+// ResourceConfig contains configuration paremeters for a resource pool
 type ResourceConfig struct {
-	ResourcePrefix string `json:"resourcePrefix,omitempty"` // optional resource prefix that will ovewrite global prefix specified in cli params
-	ResourceName   string `json:"resourceName"`             // the resource name will be added with resource prefix in K8s api
-	IsRdma         bool   // the resource support rdma
-	Selectors      struct {
-		Vendors     []string `json:"vendors,omitempty"`
-		Devices     []string `json:"devices,omitempty"`
-		Drivers     []string `json:"drivers,omitempty"`
-		PfNames     []string `json:"pfNames,omitempty"`
-		LinkTypes   []string `json:"linkTypes,omitempty"`
-		DDPProfiles []string `json:"ddpProfiles,omitempty"`
-	} `json:"selectors,omitempty"` // Whether devices have SRIOV virtual function capabilities or not
+	ResourcePrefix string           `json:"resourcePrefix,omitempty"` // optional resource prefix that will ovewrite global prefix specified in cli params
+	ResourceName   string           `json:"resourceName"`             // the resource name will be added with resource prefix in K8s api
+	IsRdma         bool             // the resource support rdma
+	DeviceType     DeviceType       `json:"deviceType,omitempty"`
+	Selectors      *json.RawMessage `json:"selectors,omitempty"`
+	DeviceFilter   DeviceFilter
```

Comment:
I wanted to do the same but the problem with that is the `DeviceFilter` will vary per resource configuration but the `DeviceProvider` will differ per "deviceType" only. We may have more than one resource configuration for same `DeviceProvider` thus the actual data in `DeviceFilter` object could be different but the data in `DeviceProvider` remain the same.

Also, the initialization of `DeviceFilter` and `DeviceProvider` are happening at different stages. The `DeviceFilter` can only be initialized during unmarsalling user configs and `DeviceProvider` during device discovery. That's why I thought it's better to keep them in separate interfaces and implementations. But the implementation within the same package of a device type.

Reply:
The reason why`DeviceFilter` varies per resource configuration is because it contains unmarshalled selectors from resource config, the rest fields in `DeviceFilter` such as `rFactory` is already contained in `DeviceProvider` structure and `isRdma` is referenced in each `ResourcePool` via unmarshalled data in `rm.configList`. In my opinion, we don't have to create a separate structure only for saving the unmarshalled selectors which looks a little over engineered.  What I was thinking is to keep using the selector structure we have so that it can be referenced during initialization of resource pool.
`DeviceFilter`, without saving unmarshalled selector config, will not be resource specific, so its function `GetFilteredDevices` can be moved to `DeviceProvider` as it will only receive selector config as argument and each `DeviceProvider` applies same selector logic.

Reply:
I sorry, I am not quite following it. Can please provide some code examples of your suggested solution?

Reply:
I wrote an example [here](https://github.com/zshi-redhat/sriov-network-device-plugin/commit/40f1cd31f09aba9b06f1ac149ad9cce905e2244c) and applied on top of this PR.

The idea is to move `GetFilteredDevices` in `DeviceProvider`:
```
type DeviceProvider interface {
	AddTargetDevices([]*ghw.PCIDevice, int) error
	GetDevices() []PciDevice
+	GetFilteredDevices(*ResourceConfig) []PciDevice
}
```

Reply:
Thanks for showing the code examples. Now it's clear!
The whole idea of moving the `Selectors` struct out from the `ResourceConfig` so that we don't have to use (or limited by) same set of selectors for all device types. Having a static struct won't allow us to have different set of selectors from using different struct. Instead we have introduced an interface instead then unmarshal the selectors using customer struct in 2nd level unmarshalling.

Reply:
I think It won't block deivce types from using device type specific selectors by keeping `Selectors` structure in `ResourceConfig` , instead it allows sharing the same selector for all types of device types. If a selector is not meaningful for a particular device type, then device provider won't handle this selector even it's configured. In enssence, one device type will only use one combination of various selectors, not two or more. The selectors are used to get filtered device only when creating resource pool during initialization, it's not necessary to carry them on to later phase where DP API is called.

Reply:
We wanted to separate out selectors for each device types so that we don't necessarily modify structure meant for one device type for another(PfNmaes, isRdma would be some examples).
The resource config is part of resource pool object regardless we use different struct per device type or the same struct as it is now. But using the approach you suggested we will also have selectors fields that may not applicable to other pool at all.
In addition to that, we will also lose some configuration validation per device type during selector unmarshalling from user input.

Reply:
> We wanted to separate out selectors for each device types so that we don't necessarily modify structure meant for one device type for another(PfNmaes, isRdma would be some examples).

We have a set of selectors defined in single structure, but it's not necessary one devcie type will use all of them, I think the device type field will be sufficient for device provider to intellegently choose which selectors to use and implememt its own selector logic as ultimately filtering device is a per device type attribute, not per resource pool. I think the discussion has drifted a little bit to focus on how selector is inputted. my original asking is more to move deviceFitler to deviceProvider. but I agree the selector structure will need further discussion and design.

> The resource config is part of resource pool object regardless we use different struct per device type or the same struct as it is now. But using the approach you suggested we will also have selectors fields that may not applicable to other pool at all.

Right, if resource pool already includes user input resourceConfig that has a full copy of selectors, it may not be needed to save another unmarshalled copy as the unmarshalling can be done at any time later. But if the goal of executing unmarshall is to examine and validate whether it has the correct format, then we can still keep the unmarshalling function, but not save the unmarshalled data.

> In addition to that, we will also lose some configuration validation per device type during selector unmarshalling from user input.

I don't see any particular validation in current code other than unmarshall raw data to a structure (if there are selectors in raw data that doesn't belong to a device type, the unmarshalling won't capture it, rather it's in device filtering function that the unexpected selector will be ignored )

Reply:
> We have a set of selectors defined in single structure, but it's not necessary one devcie type will use all of them, I think the device type field will be sufficient for device provider to intellegently choose which selectors to use and implememt its own selector logic as ultimately filtering device is a per device type attribute, not per resource pool. I think the discussion has drifted a little bit to focus on how selector is inputted. my original asking is more to move deviceFitler to deviceProvider. but I agree the selector structure will need further discussion and design.
> 

The deviceFilter works with the selector struct and it has one to one relationship with a resource pool. Whereas deviceProvider has one to many relationship with resource pool. Now, it is possible for deviceProvider also implement the DeviceFilter interface iff we use same selectors struct for all devices. This is where I'm trying to getting at - We do not want to fall into same problems using same `Selectors` struct for all devices. We want to have different set of selectors(and possibly other fields as well) per device. And modifying a selector for one device should not affect another device. Yes, in the `GetFilteredDevices()` method we could pick and choose the only fields a device maybe interested in from a common selectors struct. But I still think it's a good idea to separate out the device specific selectors and use dependency using a interface(`DeviceFilter` interface) instead of a struct(`Selectors`).

> 
> Right, if resource pool already includes user input resourceConfig that has a full copy of selectors, it may not be needed to save another unmarshalled copy as the unmarshalling can be done at any time later. But if the goal of executing unmarshall is to examine and validate whether it has the correct format, then we can still keep the unmarshalling function, but not save the unmarshalled data.
> 

With new implementation the resourceConfig no longer holds the unmarshalled selectors but a reference to DeviceFilter interface that holds the actual selectors object. Only difference the new approach to the old is we one pointer to raw bytes of selectors as placeholder to for late unmarshalling. Wherein in existing approach the resourceConfig actually holds the full Selector struct.
 
> I don't see any particular validation in current code other than unmarshall raw data to a structure (if there are selectors in raw data that doesn't belong to a device type, the unmarshalling won't capture it, rather it's in device filtering function that the unexpected selector will be ignored )

It's not obvious now, but if we had field in selector that is required(without 'omitempty' json annotation) the unmarshalling could capture this.

Reply:
Having said that, if you think it complicates things rather than adding any real benefits  then we can keep the `ResourceConfig` struct unchanged and use the common `Selectors` for all devices.

Reply:
> With new implementation the resourceConfig no longer holds the unmarshalled selectors but a reference to DeviceFilter interface that holds the actual selectors object. Only difference the new approach to the old is we one pointer to raw bytes of selectors as placeholder to for late unmarshalling. Wherein in existing approach the resourceConfig actually holds the full Selector struct.

I agree that selectors may be different per device type. I guess the argument is do we need a separate `DeviceFilter` to hold the unmarshalled data, just for use when creating resource pool? If there is a case that we were to provide different selecting logics within one device type, I'd think it makes sense. But if we were to use one selecting logic for all devices of the same device type, it might not be necessary.

> 
> > I don't see any particular validation in current code other than unmarshall raw data to a structure (if there are selectors in raw data that doesn't belong to a device type, the unmarshalling won't capture it, rather it's in device filtering function that the unexpected selector will be ignored )
> 
> It's not obvious now, but if we had field in selector that is required(without 'omitempty' json annotation) the unmarshalling could capture this.

Do we have an example of such selector that can not be omitted?

Reply:
> I agree that selectors may be different per device type. I guess the argument is do we need a separate `DeviceFilter` to hold the unmarshalled data, just for use when creating resource pool? If there is a case that we were to provide different selecting logics within one device type, I'd think it makes sense. But if we were to use one selecting logic for all devices of the same device type, it might not be necessary.
> 

I think yes, we need separate `DeviceFilter` to hold the unmarshalled data as the selector values will differ per config. And the selector logic also can be different per device type but not per config though. The DeviceProvider just holds a specific global device list based on types. It doesn't know anything about resource configurations, that is the reason it cannot implement `DeviceFilter` unless we provide config+selectors to it. The actual issue I see is the resource pool itself needs no know about the resource config and the selectors(maybe, only if we move isRdma inside a selector). 
 
> 
> Do we have an example of such selector that can not be omitted?

I don't have an explicit example in hand, but think about what if some device selector requires a flag that must be specified for that device type only but not for others? In that case a common selector won't facilitate this. 
In the original proposal the main idea was to separate out the selectors from main config so that we could have device specific selector struct.

Reply:
@zshi-redhat PTAL the recent changes.
I have moved the GetFilteredDevices into DeviceProvider but retained the Device specific selectors structs. The GetFilteredDevices() now required to get the actual selectors for resource config which we can get from a type assertion. The same techniques is used to get selectors object in the resource Pool. So, we can now get "IsRdma" from the selectors instead of the main config.

Reply:
@ahalim-intel thanks for the update! The change looks good to me. Now we allows adding device type specific selectors, but keep the same selecting logic per device type.